lib/rspec/snapshot/matchers/match_snapshot.rb

```diff
@@ -1,45 +1,78 @@
+      # ```
+      #
+      # Rules for updating snapshots (mostly borrowed from Jest)
+      #
+      # These are the conditions on when to *write* snapshots:
+      #   * The save_snapshots option is set to 'all'
+      #   * No snapshot and the save_snapshots option is 'new' (the default)
+      #
+      # These are the conditions on when *not to write* snapshots:
+      #   * The save_snapshots option is set to 'none'.
+      #   * There is a snapshot and the save_snapshots options is 'new'
       class MatchSnapShot
+        include Snapshot::LoadSnapshot
         def initialize(metadata, snapshot_name)
           @metadata = metadata
           @snapshot_name = snapshot_name
         end
 
         def matches?(actual)
           @actual = actual
-          filename = "#{@snapshot_name}.snap"
-          snap_path = File.join(snapshot_dir, filename)
-          FileUtils.mkdir_p(File.dirname(snap_path)) unless Dir.exist?(File.dirname(snap_path))
           if File.exist?(snap_path)
-            file = File.new(snap_path)
-            @expect = file.read
-            file.close
-            @actual.to_s == @expect
+            @found = true
+            # load_snapshot attempts to deserialize the way that write_snapshot serializes
+            @expect = load_snapshot(snap_path)
+            pass = compare
+            if save_config == :all && !pass
+              write_snapshot(snap_path, @actual)
+            else
+              pass
+            end
           else
-            RSpec.configuration.reporter.message "Generate #{snap_path}"
-            file = File.new(snap_path, "w+")
-            file.write(@actual)
-            file.close
-            true
+            @found = false
+            if %i[all new].include?(save_config)
+              write_snapshot(snap_path, @actual)
+            else
+              false
+            end
           end
         end
 
-
+        # TODO: more helpful snapshot diff in the failure message
         def failure_message
-          "\nexpected: #{@expect}\n     got: #{@actual}\n"
-        end
-
-        def snapshot_dir
-          if RSpec.configuration.snapshot_dir.to_s == 'relative'
-            File.dirname(@metadata[:file_path]) << "/__snapshots__"
+          if @found
+            "\nexpected: #{@expect}\n     got: #{@actual}\n"
           else
-            RSpec.configuration.snapshot_dir
+            "\nCould not find a snapshot at #{snap_path}.\n Failing " \
+              "instead of creating; save_snapshots is set to #{save_config}."
           end
         end
+
+        private def write_snapshot(snap_path, serialized_value)
+          Snapshot::Utils.write_snapshot(snap_path, serialized_value)
+          true
+        end
+
+        private def save_config
+          RSpec.configuration.save_snapshots
+        end
+
+        private def snap_path
+          Snapshot::Utils.snapshot_path(@snapshot_name)
+        end
+
+        # Todo: pass serializer
```

Comment:
It would be _awesome_ to have custom snapshot serializers (a la [jest](https://facebook.github.io/jest/docs/en/expect.html#expectaddsnapshotserializerserializer))